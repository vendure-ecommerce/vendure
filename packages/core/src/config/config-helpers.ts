import path from 'path';

import { mergeConfig } from './merge-config';
import { PartialVendureConfig, RuntimeVendureConfig } from './vendure-config';

let activeConfig: RuntimeVendureConfig;
const defaultConfigPath = path.join(__dirname, 'default-config');
/**
 * Reset the activeConfig object back to the initial default state.
 */
export function resetConfig() {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    activeConfig = require(defaultConfigPath).defaultConfig;
}

/**
 * Override the default config by merging in the supplied values. Should only be used prior to
 * bootstrapping the app.
 */
export async function setConfig(userConfig: PartialVendureConfig) {
    if (!activeConfig) {
        activeConfig = (await import(defaultConfigPath)).defaultConfig;
    }
    activeConfig = mergeConfig(activeConfig, userConfig);
}

/**
 * Ensures that the config has been loaded. This is necessary for tests which
 * do not go through the normal bootstrap process.
 */
export async function ensureConfigLoaded() {
    if (!activeConfig) {
        activeConfig = (await import(defaultConfigPath)).defaultConfig;
    }
}

/**
 * Returns the app config object. In general this function should only be
 * used before bootstrapping the app. In all other contexts, the {@link ConfigService}
 * should be used to access config settings.
 */
export function getConfig(): Readonly<RuntimeVendureConfig> {
    if (!activeConfig) {
        try {
            // eslint-disable-next-line @typescript-eslint/no-var-requires
            activeConfig = require(defaultConfigPath).defaultConfig;
        } catch (e: any) {
            // eslint-disable-next-line no-console
            console.log(
                'Error loading config. If this is a test, make sure you have called ensureConfigLoaded() before using the config.',
            );
        }
    }
    return activeConfig;
}
