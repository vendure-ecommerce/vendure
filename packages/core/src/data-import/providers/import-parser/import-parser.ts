import { Injectable } from '@nestjs/common';
import { GlobalFlag } from '@vendure/common/lib/generated-types';
import { normalizeString } from '@vendure/common/lib/normalize-string';
import { unique } from '@vendure/common/lib/unique';
import parse from 'csv-parse';
import { Stream } from 'stream';

export type BaseProductRecord = {
    name?: string;
    slug?: string;
    description?: string;
    assets?: string;
    facets?: string;
    optionGroups?: string;
    optionValues?: string;
    sku?: string;
    price?: string;
    taxCategory?: string;
    stockOnHand?: string;
    trackInventory?: string;
    variantAssets?: string;
    variantFacets?: string;
};

export type RawProductRecord = BaseProductRecord & { [customFieldName: string]: string };

export interface ParsedProductVariant {
    optionValues: string[];
    sku: string;
    price: number;
    taxCategory: string;
    stockOnHand: number;
    trackInventory: GlobalFlag;
    assetPaths: string[];
    facets: Array<{
        facet: string;
        value: string;
    }>;
    customFields: {
        [name: string]: string;
    };
}

export interface ParsedProduct {
    name: string;
    slug: string;
    description: string;
    assetPaths: string[];
    optionGroups: Array<{
        name: string;
        values: string[];
    }>;
    facets: Array<{
        facet: string;
        value: string;
    }>;
    customFields: {
        [name: string]: string;
    };
}

export interface ParsedProductWithVariants {
    product: ParsedProduct;
    variants: ParsedProductVariant[];
}

export interface ParseResult<T> {
    results: T[];
    errors: string[];
    processed: number;
}

const requiredColumns: Array<keyof BaseProductRecord> = [
    'name',
    'slug',
    'description',
    'assets',
    'facets',
    'optionGroups',
    'optionValues',
    'sku',
    'price',
    'taxCategory',
    'variantAssets',
    'variantFacets',
];

/**
 * Validates and parses CSV files into a data structure which can then be used to created new entities.
 */
@Injectable()
export class ImportParser {
    async parseProducts(input: string | Stream): Promise<ParseResult<ParsedProductWithVariants>> {
        const options: parse.Options = {
            trim: true,
            relax_column_count: true,
        };
        return new Promise<ParseResult<ParsedProductWithVariants>>((resolve, reject) => {
            let errors: string[] = [];

            if (typeof input === 'string') {
                parse(input, options, (err: any, records: string[][]) => {
                    if (err) {
                        errors = errors.concat(err);
                    }
                    if (records) {
                        const parseResult = this.processRawRecords(records);
                        errors = errors.concat(parseResult.errors);
                        resolve({ results: parseResult.results, errors, processed: parseResult.processed });
                    } else {
                        resolve({ results: [], errors, processed: 0 });
                    }
                });
            } else {
                const parser = parse(options);
                const records: string[][] = [];
                // input.on('open', () => input.pipe(parser));
                input.pipe(parser);
                parser.on('readable', () => {
                    let record;
                    // tslint:disable-next-line:no-conditional-assignment
                    while ((record = parser.read())) {
                        records.push(record);
                    }
                });
                parser.on('error', reject);
                parser.on('end', () => {
                    const parseResult = this.processRawRecords(records);
                    errors = errors.concat(parseResult.errors);
                    resolve({ results: parseResult.results, errors, processed: parseResult.processed });
                });
            }
        });
    }

    private processRawRecords(records: string[][]): ParseResult<ParsedProductWithVariants> {
        const results: ParsedProductWithVariants[] = [];
        const errors: string[] = [];
        let currentRow: ParsedProductWithVariants | undefined;
        const headerRow = records[0];
        const rest = records.slice(1);
        const totalProducts = rest.map(row => row[0]).filter(name => name.trim() !== '').length;
        const columnError = validateRequiredColumns(headerRow);
        if (columnError) {
            return { results: [], errors: [columnError], processed: 0 };
        }
        let line = 1;
        for (const record of rest) {
            line++;
            const columnCountError = validateColumnCount(headerRow, record);
            if (columnCountError) {
                errors.push(columnCountError + ` on line ${line}`);
                continue;
            }
            const r = mapRowToObject(headerRow, record);
            if (r.name) {
                if (currentRow) {
                    populateOptionGroupValues(currentRow);
                    results.push(currentRow);
                }
                currentRow = {
                    product: parseProductFromRecord(r),
                    variants: [parseVariantFromRecord(r)],
                };
            } else {
                if (currentRow) {
                    currentRow.variants.push(parseVariantFromRecord(r));
                }
            }
            const optionError = validateOptionValueCount(r, currentRow);
            if (optionError) {
                errors.push(optionError + ` on line ${line}`);
            }
        }
        if (currentRow) {
            populateOptionGroupValues(currentRow);
            results.push(currentRow);
        }
        return { results, errors, processed: totalProducts };
    }
}

function populateOptionGroupValues(currentRow: ParsedProductWithVariants) {
    const values = currentRow.variants.map(v => v.optionValues);
    currentRow.product.optionGroups.forEach((og, i) => {
        og.values = unique(values.map(v => v[i]));
    });
}

function validateRequiredColumns(r: string[]): string | undefined {
    const rowKeys = r;
    const missing: string[] = [];
    for (const col of requiredColumns) {
        if (!rowKeys.includes(col)) {
            missing.push(col);
        }
    }
    if (missing.length) {
        return `The import file is missing the following columns: ${missing.map(m => `"${m}"`).join(', ')}`;
    }
}

function validateColumnCount(columns: string[], row: string[]): string | undefined {
    if (columns.length !== row.length) {
        return `Invalid Record Length: header length is ${columns.length}, got ${row.length}`;
    }
}

function mapRowToObject(columns: string[], row: string[]): { [key: string]: string } {
    return row.reduce((obj, val, i) => {
        return { ...obj, [columns[i]]: val };
    }, {});
}

function validateOptionValueCount(
    r: BaseProductRecord,
    currentRow?: ParsedProductWithVariants,
): string | undefined {
    if (!currentRow) {
        return;
    }
    const optionValues = parseStringArray(r.optionValues);
    if (currentRow.product.optionGroups.length !== optionValues.length) {
        return `The number of optionValues must match the number of optionGroups`;
    }
}

function parseProductFromRecord(r: RawProductRecord): ParsedProduct {
    const name = parseString(r.name);
    const slug = parseString(r.slug) || normalizeString(name, '-');
    return {
        name,
        slug,
        description: parseString(r.description),
        assetPaths: parseStringArray(r.assets),
        optionGroups: parseStringArray(r.optionGroups).map(ogName => ({
            name: ogName,
            values: [],
        })),
        facets: parseStringArray(r.facets).map(pair => {
            const [facet, value] = pair.split(':');
            return { facet, value };
        }),
        customFields: parseCustomFields('product', r),
    };
}

function parseVariantFromRecord(r: RawProductRecord): ParsedProductVariant {
    return {
        optionValues: parseStringArray(r.optionValues),
        sku: parseString(r.sku),
        price: parseNumber(r.price),
        taxCategory: parseString(r.taxCategory),
        stockOnHand: parseNumber(r.stockOnHand),
        trackInventory:
            r.trackInventory == null || r.trackInventory === ''
                ? GlobalFlag.INHERIT
                : parseBoolean(r.trackInventory)
                ? GlobalFlag.TRUE
                : GlobalFlag.FALSE,
        assetPaths: parseStringArray(r.variantAssets),
        facets: parseStringArray(r.variantFacets).map(pair => {
            const [facet, value] = pair.split(':');
            return { facet, value };
        }),
        customFields: parseCustomFields('variant', r),
    };
}

function isRelationObject(value: string) {
    try {
        const parsed = JSON.parse(value);
        return parsed && parsed.hasOwnProperty('id');
    } catch(e) {
        return false;
    }
}

function parseCustomFields(prefix: 'product' | 'variant', r: RawProductRecord): { [name: string]: string } {
    return Object.entries(r)
        .filter(([key, value]) => {
            return key.indexOf(`${prefix}:`) === 0;
        })
        .reduce((output, [key, value]) => {
            const fieldName = key.replace(`${prefix}:`, '');
            return {
                ...output,
                [fieldName]: isRelationObject(value) ? JSON.parse(value) : value,
            };
        }, {});
}

function parseString(input?: string): string {
    return (input || '').trim();
}

function parseNumber(input?: string): number {
    return +(input || '').trim();
}

function parseBoolean(input?: string): boolean {
    if (input == null) {
        return false;
    }
    switch (input.toLowerCase()) {
        case 'true':
        case '1':
        case 'yes':
            return true;
        default:
            return false;
    }
}

function parseStringArray(input?: string, separator = '|'): string[] {
    return (input || '')
        .trim()
        .split(separator)
        .map(s => s.trim())
        .filter(s => s !== '');
}
