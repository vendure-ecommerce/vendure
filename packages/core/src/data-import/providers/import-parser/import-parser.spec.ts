import { LanguageCode } from '@vendure/common/lib/generated-types';
import fs from 'fs-extra';
import path from 'path';
import { beforeAll, describe, expect, it } from 'vitest';

import { ensureConfigLoaded } from '../../../config/config-helpers';
import { ConfigService } from '../../../config/config.service';

import { ImportParser } from './import-parser';

const mockConfigService = {
    defaultLanguageCode: LanguageCode.en,
    customFields: {
        Product: [
            {
                name: 'keywords',
                type: 'localeString',
                list: true,
            },
            {
                name: 'customPage',
                type: 'string',
            },
        ],
        ProductVariant: [
            {
                name: 'volumetric',
                type: 'int',
            },
        ],
    },
} as ConfigService;

describe('ImportParser', () => {
    beforeAll(async () => {
        await ensureConfigLoaded();
    });

    describe('parseProducts', () => {
        it('single product with a single variant', async () => {
            const importParser = new ImportParser(mockConfigService);

            const input = await loadTestFixture('single-product-single-variant.csv');
            const result = await importParser.parseProducts(input);

            expect(result.results).toMatchSnapshot();
        });

        it('single product with a multiple variants', async () => {
            const importParser = new ImportParser(mockConfigService);

            const input = await loadTestFixture('single-product-multiple-variants.csv');
            const result = await importParser.parseProducts(input);

            expect(result.results).toMatchSnapshot();
        });

        it('multiple products with multiple variants', async () => {
            const importParser = new ImportParser(mockConfigService);

            const input = await loadTestFixture('multiple-products-multiple-variants.csv');
            const result = await importParser.parseProducts(input);

            expect(result.results).toMatchSnapshot();
        });

        it('custom fields', async () => {
            const importParser = new ImportParser(mockConfigService);

            const input = await loadTestFixture('custom-fields.csv');
            const result = await importParser.parseProducts(input);

            expect(result.results).toMatchSnapshot();
        });

        it('works with streamed input', async () => {
            const importParser = new ImportParser(mockConfigService);

            const filename = path.join(__dirname, 'test-fixtures', 'multiple-products-multiple-variants.csv');
            const input = fs.createReadStream(filename);
            const result = await importParser.parseProducts(input);

            expect(result.results).toMatchSnapshot();
        });

        it('works with multilingual input', async () => {
            const importParser = new ImportParser(mockConfigService);

            const filename = path.join(__dirname, 'test-fixtures', 'multiple-languages.csv');
            const input = fs.createReadStream(filename);
            const result = await importParser.parseProducts(input);

            expect(result.results).toMatchSnapshot();
        });

        describe('error conditions', () => {
            it('reports errors on invalid option values', async () => {
                const importParser = new ImportParser(mockConfigService);

                const input = await loadTestFixture('invalid-option-values.csv');
                const result = await importParser.parseProducts(input);

                expect(result.errors).toEqual([
                    "The number of optionValues in column 'optionValues' must match the number of optionGroups on line 2",
                    "The number of optionValues in column 'optionValues' must match the number of optionGroups on line 3",
                    "The number of optionValues in column 'optionValues' must match the number of optionGroups on line 4",
                    "The number of optionValues in column 'optionValues' must match the number of optionGroups on line 5",
                ]);
            });

            it('reports error on ivalid columns', async () => {
                const importParser = new ImportParser(mockConfigService);

                const input = await loadTestFixture('invalid-columns.csv');
                const result = await importParser.parseProducts(input);

                expect(result.results).toEqual([]);
                expect(result.errors).toEqual([
                    'The import file is missing the following columns: "slug", "assets", "variantFacets"',
                ]);
            });

            it('reports error on ivalid row length', async () => {
                const importParser = new ImportParser(mockConfigService);

                const input = await loadTestFixture('invalid-row-length.csv');
                const result = await importParser.parseProducts(input);

                expect(result.errors).toEqual([
                    'Invalid Record Length: header length is 14, got 12 on line 3',
                    'Invalid Record Length: header length is 14, got 1 on line 4',
                ]);
                expect(result.results.length).toBe(2);
            });
        });
    });
});

function loadTestFixture(fileName: string): Promise<string> {
    return fs.readFile(path.join(__dirname, 'test-fixtures', fileName), 'utf-8');
}
