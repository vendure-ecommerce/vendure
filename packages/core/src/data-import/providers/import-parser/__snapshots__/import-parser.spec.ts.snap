// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImportParser > parseProducts > custom fields 1`] = `
[
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "size",
              "values": [
                "Half Imperial",
                "Quarter Imperial",
                "Full Imperial",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {
            "customPage": "grid-view",
            "keywords": "paper, stretch",
          },
          "description": "A great device for stretching paper.",
          "languageCode": "en",
          "name": "Perfect Paper Stretcher",
          "slug": "perfect-paper-stretcher",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 45.3,
        "sku": "PPS12",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {
              "volumetric": "243",
            },
            "languageCode": "en",
            "optionValues": [
              "Half Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 32.5,
        "sku": "PPS14",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {
              "volumetric": "344",
            },
            "languageCode": "en",
            "optionValues": [
              "Quarter Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 59.5,
        "sku": "PPSF",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {
              "volumetric": "656",
            },
            "languageCode": "en",
            "optionValues": [
              "Full Imperial",
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`ImportParser > parseProducts > multiple products with multiple variants 1`] = `
[
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "size",
              "values": [
                "Half Imperial",
                "Quarter Imperial",
                "Full Imperial",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {},
          "description": "A great device for stretching paper.",
          "languageCode": "en",
          "name": "Perfect Paper Stretcher",
          "slug": "Perfect-paper-stretcher",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 45.3,
        "sku": "PPS12",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Half Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 32.5,
        "sku": "PPS14",
        "stockOnHand": 11,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Quarter Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 59.5,
        "sku": "PPSF",
        "stockOnHand": 12,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Full Imperial",
            ],
          },
        ],
      },
    ],
  },
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [],
      "translations": [
        {
          "customFields": {},
          "description": "Mabef description",
          "languageCode": "en",
          "name": "Mabef M/02 Studio Easel",
          "slug": "mabef-m02-studio-easel",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 910.7,
        "sku": "M02",
        "stockOnHand": 13,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [],
          },
        ],
      },
    ],
  },
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "box size",
              "values": [
                "Box of 8",
                "Box of 12",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {},
          "description": "Really mega pencils",
          "languageCode": "en",
          "name": "Giotto Mega Pencils",
          "slug": "giotto-mega-pencils",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 4.16,
        "sku": "225400",
        "stockOnHand": 14,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Box of 8",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 6.24,
        "sku": "225600",
        "stockOnHand": 15,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Box of 12",
            ],
          },
        ],
      },
    ],
  },
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "size",
              "values": [
                "small",
                "large",
              ],
            },
          ],
        },
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "colour",
              "values": [
                "beige",
                "navy",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {},
          "description": "Keeps the paint off the clothes",
          "languageCode": "en",
          "name": "Artists Smock",
          "slug": "artists-smock",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10112",
        "stockOnHand": 16,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "small",
              "beige",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10113",
        "stockOnHand": 17,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "large",
              "beige",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10114",
        "stockOnHand": 18,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "small",
              "navy",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10115",
        "stockOnHand": 19,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "large",
              "navy",
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`ImportParser > parseProducts > single product with a multiple variants 1`] = `
[
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "size",
              "values": [
                "Half Imperial",
                "Quarter Imperial",
                "Full Imperial",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {},
          "description": "A great device for stretching paper.",
          "languageCode": "en",
          "name": "Perfect Paper Stretcher",
          "slug": "perfect-paper-stretcher",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 45.3,
        "sku": "PPS12",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Half Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 32.5,
        "sku": "PPS14",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Quarter Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 59.5,
        "sku": "PPSF",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Full Imperial",
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`ImportParser > parseProducts > single product with a single variant 1`] = `
[
  {
    "product": {
      "assetPaths": [
        "pps1.jpg",
        "pps2.jpg",
      ],
      "facets": [
        {
          "translations": [
            {
              "facet": "brand",
              "languageCode": "en",
              "value": "KB",
            },
          ],
        },
        {
          "translations": [
            {
              "facet": "type",
              "languageCode": "en",
              "value": "Accessory",
            },
          ],
        },
      ],
      "optionGroups": [],
      "translations": [
        {
          "customFields": {},
          "description": "A great device for stretching paper.",
          "languageCode": "en",
          "name": "Perfect Paper Stretcher",
          "slug": "perfect-paper-stretcher",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [
          {
            "translations": [
              {
                "facet": "material",
                "languageCode": "en",
                "value": "Wood",
              },
            ],
          },
        ],
        "price": 45.3,
        "sku": "PPS12",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [],
          },
        ],
      },
    ],
  },
]
`;

exports[`ImportParser > parseProducts > works with multilingual input 1`] = `
[
  {
    "product": {
      "assetPaths": [],
      "facets": [
        {
          "translations": [
            {
              "facet": "brand",
              "languageCode": "en",
              "value": "KB",
            },
            {
              "facet": "品牌",
              "languageCode": "zh_Hans",
              "value": "KB",
            },
          ],
        },
        {
          "translations": [
            {
              "facet": "type",
              "languageCode": "en",
              "value": "Accessory",
            },
            {
              "facet": "类型",
              "languageCode": "zh_Hans",
              "value": "饰品",
            },
          ],
        },
      ],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "size",
              "values": [
                "Half Imperial",
                "Quarter Imperial",
                "Full Imperial",
              ],
            },
            {
              "languageCode": "zh_Hans",
              "name": "size",
              "values": [
                "半英制",
                "四分之一英制",
                "全英制",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {
            "customPage": "grid-view",
            "keywords": "paper, stretch",
          },
          "description": "A great device for stretching paper.",
          "languageCode": "en",
          "name": "Perfect Paper Stretcher",
          "slug": "perfect-paper-stretcher",
        },
        {
          "customFields": {
            "customPage": "grid-view",
            "keywords": "纸张，拉伸",
          },
          "description": "一个用于拉伸纸张的伟大装置",
          "languageCode": "zh_Hans",
          "name": "完美的纸张拉伸器",
          "slug": "完美的纸张拉伸器",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 45.3,
        "sku": "PPS12",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {
              "volumetric": "243",
            },
            "languageCode": "en",
            "optionValues": [
              "Half Imperial",
            ],
          },
          {
            "customFields": {
              "volumetric": "243",
            },
            "languageCode": "zh_Hans",
            "optionValues": [
              "半英制",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 32.5,
        "sku": "PPS14",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {
              "volumetric": "344",
            },
            "languageCode": "en",
            "optionValues": [
              "Quarter Imperial",
            ],
          },
          {
            "customFields": {
              "volumetric": "344",
            },
            "languageCode": "zh_Hans",
            "optionValues": [
              "四分之一英制",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 59.5,
        "sku": "PPSF",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {
              "volumetric": "656",
            },
            "languageCode": "en",
            "optionValues": [
              "Full Imperial",
            ],
          },
          {
            "customFields": {
              "volumetric": "656",
            },
            "languageCode": "zh_Hans",
            "optionValues": [
              "全英制",
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`ImportParser > parseProducts > works with streamed input 1`] = `
[
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "size",
              "values": [
                "Half Imperial",
                "Quarter Imperial",
                "Full Imperial",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {},
          "description": "A great device for stretching paper.",
          "languageCode": "en",
          "name": "Perfect Paper Stretcher",
          "slug": "Perfect-paper-stretcher",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 45.3,
        "sku": "PPS12",
        "stockOnHand": 10,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Half Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 32.5,
        "sku": "PPS14",
        "stockOnHand": 11,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Quarter Imperial",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 59.5,
        "sku": "PPSF",
        "stockOnHand": 12,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Full Imperial",
            ],
          },
        ],
      },
    ],
  },
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [],
      "translations": [
        {
          "customFields": {},
          "description": "Mabef description",
          "languageCode": "en",
          "name": "Mabef M/02 Studio Easel",
          "slug": "mabef-m02-studio-easel",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 910.7,
        "sku": "M02",
        "stockOnHand": 13,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [],
          },
        ],
      },
    ],
  },
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "box size",
              "values": [
                "Box of 8",
                "Box of 12",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {},
          "description": "Really mega pencils",
          "languageCode": "en",
          "name": "Giotto Mega Pencils",
          "slug": "giotto-mega-pencils",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 4.16,
        "sku": "225400",
        "stockOnHand": 14,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Box of 8",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 6.24,
        "sku": "225600",
        "stockOnHand": 15,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "Box of 12",
            ],
          },
        ],
      },
    ],
  },
  {
    "product": {
      "assetPaths": [],
      "facets": [],
      "optionGroups": [
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "size",
              "values": [
                "small",
                "large",
              ],
            },
          ],
        },
        {
          "translations": [
            {
              "languageCode": "en",
              "name": "colour",
              "values": [
                "beige",
                "navy",
              ],
            },
          ],
        },
      ],
      "translations": [
        {
          "customFields": {},
          "description": "Keeps the paint off the clothes",
          "languageCode": "en",
          "name": "Artists Smock",
          "slug": "artists-smock",
        },
      ],
    },
    "variants": [
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10112",
        "stockOnHand": 16,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "small",
              "beige",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10113",
        "stockOnHand": 17,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "large",
              "beige",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10114",
        "stockOnHand": 18,
        "taxCategory": "standard",
        "trackInventory": "FALSE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "small",
              "navy",
            ],
          },
        ],
      },
      {
        "assetPaths": [],
        "facets": [],
        "price": 11.99,
        "sku": "10115",
        "stockOnHand": 19,
        "taxCategory": "standard",
        "trackInventory": "TRUE",
        "translations": [
          {
            "customFields": {},
            "languageCode": "en",
            "optionValues": [
              "large",
              "navy",
            ],
          },
        ],
      },
    ],
  },
]
`;
